File: examples/go/ttl-caching/document_ttl.go
package main

import (
	"bytes"
	"encoding/json"
	"fmt"
	"log"
	"net/http"
	"os"

	"github.com/joho/godotenv"
)

func getAuthToken() (string, error) {
	var baseURL = os.Getenv("API_BASE_URL")
	var apiKey = os.Getenv("API_BASE_KEY")
	authReq := map[string]string{"api_key": apiKey}
	body, _ := json.Marshal(authReq)

	resp, err := http.Post(baseURL+"/api/auth/token", "application/json", bytes.NewBuffer(body))
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()

	var result map[string]interface{}
	json.NewDecoder(resp.Body).Decode(&result)

	return result["token"].(string), nil
}

func main() {
	// Load environment variables from the .env file
	err := godotenv.Load()
	if err != nil {
		log.Fatal(fmt.Sprintf("✗ Failed to load environment variables: %v", err))
		os.Exit(1)
	}

	var baseURL = os.Getenv("API_BASE_URL")

	// Get authentication token
	token, err := getAuthToken()
	if err != nil {
		log.Fatal("Authentication failed:", err)
	}
	fmt.Println("✓ Authentication successful")

	client := &http.Client{}

	// Example 1: Insert document with TTL (1 hour)
	fmt.Println("\n=== Insert Document with TTL (1 hour) ===")
	doc1Data := map[string]interface{}{
		"key":   "session_token",
		"value": "abc123",
		"ttl":   3600,
	}
	doc1JSON, _ := json.Marshal(doc1Data)

	req, _ := http.NewRequest("POST", baseURL+"/api/insert/ttl_cache", bytes.NewBuffer(doc1JSON))
	req.Header.Set("Authorization", "Bearer "+token)
	req.Header.Set("Content-Type", "application/json")

	resp, err := client.Do(req)
	if err != nil {
		log.Fatal(err)
	}
	defer resp.Body.Close()

	var doc1Result map[string]interface{}
	json.NewDecoder(resp.Body).Decode(&doc1Result)
	fmt.Printf("✓ Inserted document: %v\n", doc1Result["id"])

	// Example 2: Insert with shorter TTL (integer seconds)
	fmt.Println("\n=== Insert Document with TTL (5 minutes - integer) ===")
	doc2Data := map[string]interface{}{
		"key":   "temp_data",
		"value": map[string]bool{"important": true},
		"ttl":   300,
	}
	doc2JSON, _ := json.Marshal(doc2Data)

	req, _ = http.NewRequest("POST", baseURL+"/api/insert/ttl_cache", bytes.NewBuffer(doc2JSON))
	req.Header.Set("Authorization", "Bearer "+token)
	req.Header.Set("Content-Type", "application/json")

	resp, err = client.Do(req)
	if err != nil {
		log.Fatal(err)
	}
	defer resp.Body.Close()

	var doc2Result map[string]interface{}
	json.NewDecoder(resp.Body).Decode(&doc2Result)
	fmt.Printf("✓ Inserted document: %v\n", doc2Result["id"])

	// Example 3: Insert with duration string format
	fmt.Println("\n=== Insert Document with TTL (30 minutes - duration string) ===")
	doc3Data := map[string]interface{}{
		"key":   "duration_test",
		"value": "testing duration strings",
		"ttl":   "30m",
	}
	doc3JSON, _ := json.Marshal(doc3Data)

	req, _ = http.NewRequest("POST", baseURL+"/api/insert/ttl_cache", bytes.NewBuffer(doc3JSON))
	req.Header.Set("Authorization", "Bearer "+token)
	req.Header.Set("Content-Type", "application/json")

	resp, err = client.Do(req)
	if err != nil {
		log.Fatal(err)
	}
	defer resp.Body.Close()

	var doc3Result map[string]interface{}
	json.NewDecoder(resp.Body).Decode(&doc3Result)
	fmt.Printf("✓ Inserted document with duration string TTL: %v\n", doc3Result["id"])

	// Example 4: Query documents
	fmt.Println("\n=== Query Documents ===")
	queryData := map[string]interface{}{"limit": 10}
	queryJSON, _ := json.Marshal(queryData)

	req, _ = http.NewRequest("POST", baseURL+"/api/find/ttl_cache", bytes.NewBuffer(queryJSON))
	req.Header.Set("Authorization", "Bearer "+token)
	req.Header.Set("Content-Type", "application/json")

	resp, err = client.Do(req)
	if err != nil {
		log.Fatal(err)
	}
	defer resp.Body.Close()

	var docs []interface{}
	json.NewDecoder(resp.Body).Decode(&docs)
	fmt.Printf("✓ Found %d documents with TTL\n", len(docs))

	// Example 5: Update document
	fmt.Println("\n=== Update Document ===")
	updateData := map[string]interface{}{"value": "updated_value"}
	updateJSON, _ := json.Marshal(updateData)

	req, _ = http.NewRequest("PUT", fmt.Sprintf("%s/api/update/ttl_cache/%v", baseURL, doc1Result["id"]), bytes.NewBuffer(updateJSON))
	req.Header.Set("Authorization", "Bearer "+token)
	req.Header.Set("Content-Type", "application/json")

	resp, err = client.Do(req)
	if err != nil {
		log.Fatal(err)
	}
	defer resp.Body.Close()
	fmt.Println("✓ Updated document")

	// Example 6: Delete document
	fmt.Println("\n=== Delete Document ===")
	if doc1Result["id"] != nil {
		req, _ = http.NewRequest("DELETE", fmt.Sprintf("%s/api/delete/ttl_cache/%v", baseURL, doc1Result["id"]), nil)
		req.Header.Set("Authorization", "Bearer "+token)

		resp, err = client.Do(req)
		if err != nil {
			log.Fatal(err)
		}
		defer resp.Body.Close()
		fmt.Println("✓ Deleted document")
	} else {
		fmt.Println("✓ Skipped delete (no ID returned)")
	}

	fmt.Println("\n✓ All document TTL operations completed successfully")
}

File: examples/rust/examples/document_ttl.rs
use reqwest::Client;
use serde_json::{json, Value};
use std::env;
use std::time::Duration;

const COLLECTION: &str = "ttl_cache_example_rs";

async fn delete_collection(
    client: &Client,
    base_url: &str,
    token: &str,
) -> Result<(), Box<dyn std::error::Error>> {
    let response = client
        .delete(format!("{base_url}/api/collections/{COLLECTION}"))
        .bearer_auth(token)
        .send()
        .await?;
    if response.status() != reqwest::StatusCode::NOT_FOUND {
        response.error_for_status()?;
    }
    Ok(())
}

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    println!("=== ekoDB Document TTL Example (Rust) ===");
    dotenv::dotenv().ok();

    let base_url = env::var("API_BASE_URL").unwrap_or_else(|_| "http://localhost:8080".to_string());
    let api_key =
        env::var("API_BASE_KEY").unwrap_or_else(|_| "a-test-api-key-from-ekodb".to_string());

    // Get authentication token
    let client = Client::builder().timeout(Duration::from_secs(30)).build()?;
    let auth_response: Value = client
        .post(&format!("{}/api/auth/token", base_url))
        .json(&json!({"api_key": api_key}))
        .send()
        .await?
        .json()
        .await?;

    let token = auth_response["token"].as_str().unwrap();
    println!("✓ Authentication successful");

    delete_collection(&client, &base_url, token).await?;
    let operation_result = run_operations(&client, &base_url, token).await;
    let cleanup_result = delete_collection(&client, &base_url, token).await;
    match (operation_result, cleanup_result) {
        (Err(primary), Err(cleanup)) => {
            return Err(format!("{primary}; cleanup also failed: {cleanup}").into());
        }
        (Err(primary), Ok(())) => return Err(primary),
        (Ok(()), Err(cleanup)) => return Err(cleanup),
        (Ok(()), Ok(())) => {}
    }

    println!("\n✓ All document TTL operations completed successfully");
    Ok(())
}

async fn run_operations(
    client: &Client,
    base_url: &str,
    token: &str,
) -> Result<(), Box<dyn std::error::Error>> {
    // Example 1: Insert document with TTL
    println!("\n=== Insert Document with TTL (1 hour) ===");
    let doc1: Value = client
        .post(format!("{base_url}/api/insert/{COLLECTION}"))
        .header("Authorization", format!("Bearer {}", token))
        .json(&json!({
            "key": "session_token",
            "value": "abc123",
            "ttl": 3600
        }))
        .send()
        .await?
        .json()
        .await?;
    let doc1_id = doc1["id"].as_str().ok_or("insert response omitted id")?;
    println!("✓ Inserted document: {doc1_id}");

    // Example 2: Insert with shorter TTL (integer seconds)
    println!("\n=== Insert Document with TTL (5 minutes - integer) ===");
    let doc2: Value = client
        .post(format!("{base_url}/api/insert/{COLLECTION}"))
        .header("Authorization", format!("Bearer {}", token))
        .json(&json!({
            "key": "temp_data",
            "value": "short-lived data",
            "ttl": 300
        }))
        .send()
        .await?
        .json()
        .await?;
    let doc2_id = doc2["id"].as_str().ok_or("insert response omitted id")?;
    println!("✓ Inserted document: {doc2_id}");

    // Example 3: Insert with duration string format
    println!("\n=== Insert Document with TTL (30 minutes - duration string) ===");
    let doc3: Value = client
        .post(format!("{base_url}/api/insert/{COLLECTION}"))
        .header("Authorization", format!("Bearer {}", token))
        .json(&json!({
            "key": "duration_test",
            "value": "testing duration strings",
            "ttl": "30m"
        }))
        .send()
        .await?
        .json()
        .await?;
    let doc3_id = doc3["id"].as_str().ok_or("insert response omitted id")?;
    println!("✓ Inserted document with duration string TTL: {doc3_id}");

    // Example 3: Query documents
    println!("\n=== Query Documents ===");
    let docs: Vec<Value> = client
        .post(format!("{base_url}/api/find/{COLLECTION}"))
        .header("Authorization", format!("Bearer {}", token))
        .json(&json!({"limit": 10}))
        .send()
        .await?
        .json()
        .await?;
    println!("✓ Found {} documents with TTL", docs.len());

    // Example 4: Update document
    println!("\n=== Update Document ===");
    client
        .put(format!("{base_url}/api/update/{COLLECTION}/{doc1_id}"))
        .header("Authorization", format!("Bearer {}", token))
        .json(&json!({"value": "updated_value"}))
        .send()
        .await?
        .error_for_status()?;
    println!("✓ Updated document");

    // Example 6: Delete document
    println!("\n=== Delete Document ===");
    client
        .delete(format!("{base_url}/api/delete/{COLLECTION}/{doc1_id}"))
        .header("Authorization", format!("Bearer {}", token))
        .send()
        .await?
        .error_for_status()?;
    println!("✓ Deleted document");

    // doc2_id and doc3_id are validated above; the owned collection cleanup
    // removes them together even if a preceding operation fails.
    let _ = (doc2_id, doc3_id);
    Ok(())
}
